src/Container/ContainerLogItem.php

```diff
@@ -9,33 +9,43 @@
 use ReflectionNamedType;
 use ReflectionParameter;
 use ReflectionUnionType;
+use Reflector;
 
 final readonly class ContainerLogItem
 {
     public function __construct(
         public string $id,
-        public ?ReflectionClass $class = null,
-        public ?ReflectionMethod $method = null,
-        public ?ReflectionParameter $parameter = null,
+        public ?Reflector $subject = null
     ) {
     }
 
     public function __toString(): string
     {
-        return match (true) {
-            $this->parameter !== null => $this->parameterAsString(),
-            $this->method !== null => "{$this->method->getDeclaringClass()->getName()}::{$this->method->getName()}()",
-            $this->class !== null => "{$this->class->getName()}",
+        $subjectType = $this->subject ? $this->subject::class : null;
+
+        return match ($subjectType) {
+            ReflectionParameter::class => $this->formatParameter(),
+            ReflectionMethod::class => $this->formatMethod(),
+            ReflectionClass::class => $this->formatClass(),
             default => $this->id,
         };
     }
 
-    private function parameterAsString(): string
+    private function formatParameter(): string
     {
-        $type = $this->shortTypeName($this->parameter->getType());
-        $class = $this->shortTypeName($this->parameter->getDeclaringClass());
+        $type = $this->shortTypeName($this->subject->getType());
+
+        return " [unresolved parameter: {$type} \${$this->subject->getName()}]";
```

Comment:
We do need the location as well though, because otherwise it kind of becomes useless. I hard coded `__construct` because apparently `Parameter` has no reference to its method. We could improve that, but we can't keep it like this.

Also, de the `[ ]` brackets add value?

Reply:
The "step" keeps the context of the class / method / parameter so that we can output this information. Brackets just add visual appeal. 🙃